File: src/main/java/cz/linkedlist/TileListingService.java
package cz.linkedlist;

import java.time.LocalDate;
import java.util.Collection;
import java.util.List;
import java.util.Set;

/**
 * @author Martin Macko <https://github.com/LinkedList>.
 */
public interface TileListingService {

    List<String> getFolderContents(TileSet tileSet);

    boolean exists(TileSet tileSet);

    Set<Integer> getYears(UTMCode code);

    Set<Integer> getMonths(UTMCode code, int year);

    Set<Integer> getDays(UTMCode code, int year, int month);

    Set<Integer> getDataSets(UTMCode code, int year, int month, int day);

    List<LocalDate> availableDates(UTMCode utmCode);

    Collection<LocalDate> availableDatesAfter(UTMCode utmCode, LocalDate date);

    default String stripPrefixAnsSlash(String prefix, String value) {
        String strippedPrefix = value.substring(prefix.length());
        return strippedPrefix.substring(0, strippedPrefix.length() - 1);
    }
}

File: src/main/kotlin/cz/linkedlist/TileListingService.kt
package cz.linkedlist

import java.time.LocalDate

/**
 * @author Martin Macko <https:></https:>//github.com/LinkedList>.
 */
interface TileListingService {

    fun getFolderContents(tileSet: TileSet): List<String>

    fun exists(tileSet: TileSet): Boolean

    fun getYears(code: UTMCode): Set<Int>

    fun getMonths(code: UTMCode, year: Int): Set<Int>

    fun getDays(code: UTMCode, year: Int, month: Int): Set<Int>

    fun getDataSets(code: UTMCode, year: Int, month: Int, day: Int): Set<Int>

    fun availableDates(utmCode: UTMCode): List<LocalDate>

    fun availableDatesAfter(utmCode: UTMCode, date: LocalDate): Collection<LocalDate>

    fun stripPrefixAnsSlash(prefix: String, value: String): String {
        val strippedPrefix = value.substring(prefix.length)
        return strippedPrefix.substring(0, strippedPrefix.length - 1)
    }
}
